agent/consul/state/catalog.go

```diff
@@ -3491,13 +3502,36 @@ func terminatingConfigGatewayServices(
 			CertFile:    svc.CertFile,
 			CAFile:      svc.CAFile,
 			SNI:         svc.SNI,
+			Kind:        kind,
 		}
 
 		gatewayServices = append(gatewayServices, mapping)
 	}
 	return false, gatewayServices, nil
 }
 
+func GatewayServiceKind(tx ReadTxn, name string, entMeta *acl.EnterpriseMeta) (structs.GatewayServiceKind, error) {
+	serviceIter, err := tx.First(tableServices, indexService, Query{
+		Value:          name,
+		EnterpriseMeta: *entMeta,
+	})
+	if err != nil {
+		return structs.GatewayservicekindUnknown, err
+	}
+	if serviceIter != nil {
+		return structs.GatewayservicekindService, err
+	}
+
+	_, entry, err := configEntryTxn(tx, nil, structs.ServiceDefaults, name, entMeta)
+	if err != nil {
+		return structs.GatewayservicekindUnknown, err
+	}
+	if entry != nil {
+		return structs.GatewayservicekindDestination, nil
```

Comment:
Shouldn't this deep-inspect the config entry to see if it's got destination-y bits set? It's valid to have no registered instances of a service and a `service-defaults` for something that _will_ be in the catalog.

Reply:
yes right, good catch! Also that point me that I'm missing a test related a normal service-defaults.